app/play.py

```diff
@@ -212,9 +212,24 @@ async def get_completion(input_text, output_file, base_url, temperature, max_tok
 
         answer = []
         print("\n" + "*" * 100)
-        for chunk in runnable.stream({"input_text": input_text}):
-            print(chunk.content, end="", flush=True)
-            answer.append(chunk.content)
+        
+        # Check for the token_usage flag if it is present or not.
+        # If present, retrieve the output and input tokens used for the completion.
+        # Making two identical loops helps us prevent the IF checks in the loop if the token_usage flag is not used.
+        if token_usage:
```

Comment:
I think this can go inside for loop and  if argument --token-usage is true we get completion_token and prompt token otherwise just not get those token

Reply:
```python3
if token_usage and chunk.usage_metadata:
        completion_tokens = chunk.usage_metadata.get('output_tokens')
        prompt_tokens = chunk.usage_metadata.get('input_tokens')
```

Reply:
To clarify, you want me to change this:
```python
if token_usage:
    for chunk in runnable.stream({"input_text": input_text}):
        print(chunk.content, end="", flush=True)
        answer.append(chunk.content)

        # Check for the attribute usage_metadata in the chunk.
        # Retrieve the output and input tokens if available.
        if chunk.usage_metadata:
            # usage_metadata = {'output_tokens': number, 'input_tokens': number, 'total_tokens': number}
            completion_tokens = chunk.usage_metadata.get('output_tokens')
            prompt_tokens = chunk.usage_metadata.get('input_tokens')
else:
    for chunk in runnable.stream({"input_text": input_text}):
        print(chunk.content, end="", flush=True)
        answer.append(chunk.content)
```

into a single `for` loop?
If that is the case, YES we can do that, however, I wrote the code this way to prevent some overhead.

If we only use one loop, that means that we will always do an `if` check every iteration of the loop even if `token_usage` is not used. This means, we are checking for this logic every loop even if we do not need it. Splitting this code means that you do not need to do this check every loop if `token_usage` flag was not used!

Let me know what you think!

Reply:
You are correct, in term of performance your code will do better here. 

Thanks I will accept it.

Reply:
@cleobnvntra  Can you rebase the repo again the test file got lint error